openmp/openmp_cluster.cpp

```diff
@@ -0,0 +1,120 @@
+#include "../lib/types.cpp"
+#include "../sequential/lib/helpers.cpp"
+#include <float.h>
+#include <iostream>
+#include <omp.h>
+using namespace std;
+
+#define FLOAT_MAX 3.402823e+38
+
+// cluster assignment using randomization
+int * init_cluster_assignment(int k, int size, int * cluster_size){
+  for (int i=0; i<k; i++)
+    cluster_size[i] = 0;
+  int group = 0;
+
+  int * ret;
+  ret = new int[size];
+  srand(static_cast<unsigned int>(clock()));
+  for (int i=0; i<size; i++){
+    group = (int) rand() % k;
+    ret[i] = group;
+    cluster_size[group] += 1;
+  }
+  return ret;
+}
+
+bool update_clusters(int k, float ** cluster, const int * cluster_assignment, const cluster_dataset &data, const int * cluster_size){
+  bool did_change = 0;
+
+  float ** temp;
+  temp = new float* [k];
+  for (int i=0; i<k; i++)
+    temp[i] = new float[data.dimensions];
+
+  int i;
+
+#pragma omp parallel for private(i)
+  for (i=0; i<data.size; i++){
+    for (int j=0; j<data.dimensions; j++){
+      temp[cluster_assignment[i]][j] = data.features[i][j];
```

Comment:
should this be `+=` instead of `=`?

Reply:
Oops! Yeah you're right I was misunderstanding this loop. It can't be parallelized because of a race condition anyway.